pkg/operator/controller/configurable-route/controller.go

```diff
@@ -0,0 +1,393 @@
+	errors := []error{}
+
+	existingHashes := map[string]struct{}{}
+	for _, cr := range componentRoutes {
+		existingHashes[cr.Hash] = struct{}{}
+	}
+
+	roleList := &rbacv1.RoleList{}
+	r.cache.List(context.TODO(), roleList, allComponentRouteResources()...)
+	for _, item := range roleList.Items {
+		expectedHash, ok := item.GetLabels()[componentRouteHashAnnotation]
+		if !ok {
+			errors = append(errors, fmt.Errorf("Unable to find componentRoute hash label on roleBinding %s/%s", item.Namespace, item.Name))
+			continue
+		}
+
+		if _, ok := existingHashes[expectedHash]; !ok {
+			log.Info("deleting role", "name", item.GetName(), "namespace", item.GetNamespace())
+			if err := r.client.Delete(context.TODO(), &item); err != nil && !apierrors.IsNotFound(err) {
+				errors = append(errors, err)
+			}
+		}
+	}
+
+	return errors
+}
+
+func (r *reconciler) deleteOrphanedRoleBindings(componentRoutes []aggregatedComponentRoute) []error {
+	errors := []error{}
+
+	existingHashes := map[string]struct{}{}
+	for _, cr := range componentRoutes {
+		existingHashes[cr.Hash] = struct{}{}
+	}
+
+	roleBindingList := &rbacv1.RoleBindingList{}
+	r.cache.List(context.TODO(), roleBindingList, allComponentRouteResources()...)
+	for _, item := range roleBindingList.Items {
+		expectedHash, ok := item.GetLabels()[componentRouteHashAnnotation]
+		if !ok {
+			errors = append(errors, fmt.Errorf("Unable to find componentRoute hash label on roleBinding %s/%s", item.Namespace, item.Name))
+			continue
+		}
+
+		if _, ok := existingHashes[expectedHash]; !ok {
+			log.Info("deleting roleBinding", "name", item.GetName(), "namespace", item.GetNamespace())
+			if err := r.client.Delete(context.TODO(), &item); err != nil && !apierrors.IsNotFound(err) {
+				errors = append(errors, err)
+			}
+		}
+	}
+
+	return errors
+}
+
+func (r *reconciler) ensureServiceCertKeyPairSecretRole(owner metav1.Object, componentRoute aggregatedComponentRoute) (string, error) {
+	role := &rbacv1.Role{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: componentRoute.Name + "-",
+			Namespace:    r.config.SecretNamespace,
+			Labels: map[string]string{
+				componentRouteHashAnnotation: componentRoute.Hash,
+			},
+		},
+		Rules: []rbacv1.PolicyRule{
+			{
+				Verbs:         []string{"get", "list", "watch"},
+				APIGroups:     []string{""},
+				Resources:     []string{"secrets"},
+				ResourceNames: []string{componentRoute.ServingCertificateName},
+			},
+		},
+	}
+
+	roleList := &rbacv1.RoleList{}
+	if err := r.cache.List(context.TODO(), roleList, componentRouteResources(componentRoute)...); err != nil {
+		return "", err
+	}
+
+	if len(roleList.Items) == 0 {
+		if err := r.client.Create(context.TODO(), role); err != nil {
+			return "", err
+		}
+	} else {
+		role.Name = roleList.Items[0].Name
+		role.GenerateName = ""
+		if _, _, err := resourceapply.ApplyRole(r.kclient.RbacV1(), r.eventRecorder, role); err != nil {
+			return "", err
+		}
+	}
+
+	return role.GetName(), nil
```

Comment:
Do you need to do another get to get the generated name?

Reply:
Nope, the name is updated in the create call. I do however need to do a call later to get the UID of the role so it can be used on the roleBinding's ownerReference.